horovod/common/operations.cc

```diff
@@ -1282,14 +1294,29 @@ bool RunLoopOnce(HorovodGlobalState& state, MPIContext& ctx, bool is_coordinator
 
   // Flag indicating that the background thread should shut down.
   bool should_shut_down = state.shut_down;
-  if (is_coordinator) {
+  if (is_coordinator && state.perform_stall_check) {
     // Check for stalled tensors.
-    if (state.perform_stall_check &&
-        std::chrono::steady_clock::now() - state.last_stall_check >
+    if (std::chrono::steady_clock::now() - state.last_stall_check >
             std::chrono::seconds(state.stall_warning_time_seconds)) {
       should_shut_down |= CheckForStalledTensors(state, ctx);
       state.last_stall_check = std::chrono::steady_clock::now();
     }
+
+    // Check if response cache should be cleared to trigger stall warnings.
+    if (state.response_cache.capacity() > 0) {
+        if (!state.cached_messages_pending) {
+          // If no outstanding cached messages, reset perform operation clock.
+          state.last_perform_operation = std::chrono::steady_clock::now();
+        } else if (std::chrono::steady_clock::now() - state.last_perform_operation >
```

Comment:
Is there any way for us to track that there's a tensor that's been sitting in the queue for more than X time rather than tracking perform operations, add that as a bit flag and OR it?  It's possible that there has been no tensors to reduce for a while (running a large validation), or (less likely): there are two application threads, and one thread keeps reducing something, while other thread is stalled.

Reply:
The `cached_messages_pending` is set from an additional bit status flag I added to the cache coordinator (`cached_in_queue`) that indicates if any worker has a cached tensor pending (see line here: https://github.com/horovod/horovod/pull/902/files#diff-98391bc0c5b3bf87519e745b4ec3b9fdR796). So in the case that no tensors are reduced for a while, this flag is false and the stall message logic isn't triggered (and the `last_perform_operation` time point is updated so that once another cache messaged shows up, the stall warning isn't immediately triggered).

I considered keep track of how long the `cached_messages_pending` flag remains true and if that exceeds a threshold, trigger the stall message. The issue with this though is that this flag does cannot discern stale tensors, so in the case that you have 60 seconds of continuous cycles with tensors to reduce, it would view that as a stall condition. This is why I based the stall timing on the time since the last perform operation.

Reply:
Do individual workers see that they have pending tensors?  Maybe we could change `cached_messages_pending` to mean that the tensor is pending for 60 seconds?

Reply:
Hmm these are good points. I can keep track of individual cached tensors when they are replaced to the state queue here: https://github.com/horovod/horovod/pull/902/files#diff-98391bc0c5b3bf87519e745b4ec3b9fdR1345
I think in this way, I should be able to check if a specific tensor has been pending for >60 seconds on any worker and if so, trigger an invalidation of that tensor entry to get the stall message rather than a full cache flush. I'll try that instead.

Reply:
Okay, just pushed an alternative version of the stall message handling for cached messages. I just keep a map of pending cached tensors on every worker and the initial time they are queued. If a worker sees a tensor listed in that map for greater than 60 seconds, it triggers an invalidation of that cache entry. Entries are removed from the map when they are forwarded to be processed either by RunBypass or the normal path.